src/usr/bin/google_set_multiqueue

```diff
@@ -45,48 +45,67 @@ function set_channels() {
 
 function set_irq_range() {
   local -r nic="$1"
-  local core="$2"
+  local bind_cores_begin="$2"
 
   # The user may not have this $nic configured on their VM, if not, just skip
   # it, no need to error out.
   if [ ! -d "/sys/class/net/"$nic"/device" ]; then
     return;
   fi
 
+
   # We count the number of rx queues and assume number of rx queues == tx
-  # queues. Currently the GVE configuration at boot is 16 rx + 16 tx.
+  # queues. The number of queues shown in the sysfs stands for the initial
+  # queues while the number of IRQs stands for the max queues. The number of
+  # initial queues should be always less than or equal to that of the max
+  # queues.
+  core=$bind_cores_begin
+  num_irqs=$(( $(ls /sys/class/net/"$nic"/device/msi_irqs | wc -l) / 2 ))
   num_q=$(ls -1 /sys/class/net/"$nic"/queues/ | grep rx | wc -l)
 
   echo "Setting irq binding for "$nic" to core [$core - $((core + num_q - 1))] ..."
 
   irqs=($(ls /sys/class/net/"$nic"/device/msi_irqs | sort -g))
-  for ((queue = 0; queue < "$num_q"; queue++)); do
-    tx_irq=${irqs[$queue]}
-    rx_irq=${irqs[$((queue + num_q))]}
+  for ((irq = 0; irq < "$num_irqs"; irq++)); do
+    tx_irq=${irqs[$irq]}
+    rx_irq=${irqs[$((irq + num_irqs))]}
+
+    # Only allocate $num_q cores to the IRQs and queues. If the number of IRQs
+    # is more than that of queues, the CPUs will be wrapped around.
+    core=$(( bind_cores_begin + irq % num_q ))
 
     # this is GVE's TX irq. See gve_tx_idx_to_ntfy().
     echo "$core" > /proc/irq/"$tx_irq"/smp_affinity_list
 
     # this is GVE's RX irq. See gve_rx_idx_to_ntfy().
     echo "$core" > /proc/irq/"$rx_irq"/smp_affinity_list
 
-    # XPS (Transmit Packet Steering) allows a core to decide which queue to
-    # select if its mask is found in one of the queue's xps_cpus
-    cp /proc/irq/"$tx_irq"/smp_affinity /sys/class/net/"$nic"/queues/tx-"$queue"/xps_cpus
-
-    echo -en "$nic:q-$queue: \ttx: irq $tx_irq bind to $core \trx: irq $rx_irq bind to $core"
-    echo -e " \txps_cpus bind to $(cat /sys/class/net/"$nic"/queues/tx-"$queue"/xps_cpus)"
+    # Check if the queue exists at present because the number of IRQs equals
+    # the max number of queues allocated and could be greater than the current
+    # number of queues.
+    tx_queue=/sys/class/net/"$nic"/queues/tx-"$irq"
+    if ls $tx_queue 1> /dev/null 2>&1; then
```

Comment:
when is this condition expected to be false?

Reply:
I'm assuming this block is intended for A3+ cases, right?

Reply:
The previous logic was assuming the `/sys/class/net/"$nic"/queues/tx-"$queue"` always exists. This works for A3 because the number of initial queues(the queue number when the VM is booted) equals the max number of queues. Here, `num_q` is the number of initial queues while `num_irqs` is the max number of queues. They are both 16 for A3.

However, this is not true for A3+ since we reduced the initial queues for fabric NICs(eth1 - eth8) from 16 to 8 but keep the max number at 16.  Since we still want to assign the SMP affinity to all the IRQs but only assign it to the initial queues, we need to check if the queue exists in sysfs.

Reply:
sounds good to me. Thanks.